### ISSUE TITLE
Architectural decisions

### ISSUE THREAD

**Issue by danielfc** [kalnee/bearinmind#1]
Following are the options we have to create the web app. Mostly JavaScript and Java frameworks.

**Front end**
* [ ] AngularJS
* [ ] Ember.js
* [ ] Meteor
* [ ] React.js

**Back end**
* [ ] Node/Sails.js
* [ ] Node/Meteor
* [ ] Node/Express
* [ ] Node/Hapi.js
* [ ] Java EE
* [ ] Java/Dropwizard
* [ ] PHP/Laravel

**Database**
* [ ] MongoDB
* [ ] PostgreSQL
* [ ] MySQL
* [ ] RethinkDB

**Comment by danielfc** [comment]
I'd vote for `Sail.js`+`Ember.js` because we never worked with both of them and they seem promising. However, `Meteor` is growing fast and the fact that the project got a funding of millions of dollars makes it interesting. I personally prefer an architecture with a clear separation between front and back end but I also like to learn a technology that offers every layer. Do your research and when you get some time, answer the question.

**Comment by danielfc** [comment]
This guy advises to think in design patterns rather than frameworks.Take a look when you have some time.

https://www.youtube.com/watch?v=WOVmr6CjgNw&index=4&list=PLUPi8Qj7uZ3QpsamMg8NvqI9QWv3bK974

**Comment by danielfc** [comment]
In previous conversation, we defined that we will be using `React.js + Flux + Node.js (Sails.js)` as the stack of our application.

*Reasons*

1. `React.js` is the second most popular JavaScript framework, right after Angular, and it provides a better performance.
2. `Node.js/Sails.js` in the backend because it is an evolving technology being adopted by countless companies around the world.

**Comment by danielfc** [comment]
Look how things change along the way, I wouldn't even consider Ember.js anymore. Sail.js has faded away, I haven't heard anything about it for months. React is very popular with a high demand, but I think using it with all the libs it requires (redux, etc..) is not worth. AngularJS 2 is coming and I think it will rule in terms of JS frameworks in next few years. Aurora is another one to keep an eye on.

Today, I'd say we should definitely go with Angular 2 with a MEAN architecture or with a Java Backend (Dropwizard (easier, simpler, better performance) or Spring Boot (more popular, high demand)).

@joaomcar I'd like to "hear" your thought on the subject.

**Comment by joaomcar** [comment]
I think we should use Angular 2 in the frontend and Java in the backend. Although we have some experience in node, we are proficient in Java so I think it's a more reasonable choice. I would pick Spring Boot because of the highest demand. What do you think @danielfc?

**Comment by danielfc** [comment]
I totally agree with Angular 2 in the FE side. As for the BE, it's something to think about -- we won't have the first real opportunity with Node.js if we don't use it in side projects like this one. Java is something we have a lot of experience and might not be a good choice here. On the other hand, on a production scenario, I'd definitely choose Java for the BE, being DW or Spring Boot the best options. It's more of a decision between open our range of technologies or to dig deep in the Java environment.

**Comment by joaomcar** [comment]
Ok, so let's open our range and use Node with ES6 in the backend. What database should we use in this case?

**Comment by danielfc** [comment]
in this case, I'd vote for mean.io, so mongo, or mysql using something like https://github.com/mysqljs/mysql

**Comment by joaomcar** [comment]
I don't know what benefits we'd have by using Mongo but if we want to open our range it would be the right choice.
